### ISSUE TITLE
PCLVisualizer::addPointCloudNormals does not account for cloud pose

### ISSUE THREAD

**Issue by ahundt** [PointCloudLibrary/pcl#1475]
[PCLVisualizer::addPointCloudNormals](https://github.com/PointCloudLibrary/pcl/blob/master/visualization/include/pcl/visualization/impl/pcl_visualizer.hpp#L722) does not account for the position/orientation of the cloud, while the equivalent add cloud/update cloud in the same class does. This leads to the following problem:

<img width="960" alt="cloud and normal display failure" src="https://cloud.githubusercontent.com/assets/55744/11764497/cbb90f2e-a100-11e5-9dec-63afaffd49a9.png">

**Comment by andersgb1** [comment]
This looks weird! There must be something more going on in your code. The function you link to just inserts lines at the positions of the points of the first argument (cloud). Minimal code example?

**Comment by ahundt** [comment]
@andersgb1 did you look at the file I linked to? Based on the pcl code, [addPointCloud()](https://github.com/PointCloudLibrary/pcl/blob/master/visualization/include/pcl/visualization/impl/pcl_visualizer.hpp#L118) clearly uses `cloud->sensor_origin_, cloud->sensor_orientation_`, while the corresponding code in [PCLVisualizer::addPointCloudNormals](https://github.com/PointCloudLibrary/pcl/blob/master/visualization/include/pcl/visualization/impl/pcl_visualizer.hpp#L722) does not.

Since I've set both the normal and regular point clouds to have a common origin pose in space, they should be rendered at the same position. However, since the rendering code for normals never even accesses  `cloud->sensor_origin_` or `cloud->sensor_orientation_`, I can conclude it is not used or handled correctly, and thus due to a bug in PCL. Hopefully that explains why the issue is in PCL. 

For a minimal example, I guess any code where you load up a .ply with points and normals where you change the orientation and position of the cloud then display both will show the problem.

**Comment by andersgb1** [comment]
Hi @ahundt,

Yes I know the function. You basically gave the information I was after in my first comment - that you have a non-identity transform in your point cloud header.

However, my other comment still remains: how does it occur that you are getting normal lines at different positions than in your **cloud** when **addPointCloudNormals()** does not even use the XYZ information in the **normals** to place these lines? Only the XYZ information in the **cloud** is used for that. This is why I'm just asking for a minimal working example (with data) such that any PCL developer can reproduce your error and fix it (just like we ask everyone else here to do).

**Comment by SergioRAgostinho** [comment]
>  Only the XYZ information in the cloud is used for that. 

It still needs to apply the sensor transformation to the normals' origins (points) and to the normals' directions, for consistent viewing with ```addPointCloud()```. 

Nevertheless, minimal code example and if possible a patch would be appreciated :smile:.  

@VictorLamoine if you confirmed the bug, can you post the code example you used.

**Comment by andersgb1** [comment]
@SergioRAgostinho true, but my point is that the normals' origins (points) are disregarded in the linked-to function. So principally it only needs to rotate the normal orientations - and of course make a full transformation of the cloud points.

**Comment by SergioRAgostinho** [comment]
I'm not an expert in VTK, but based on what I'm seeing [here](https://github.com/PointCloudLibrary/pcl/blob/master/visualization/include/pcl/visualization/impl/pcl_visualizer.hpp#L767-L769) and [here](https://github.com/PointCloudLibrary/pcl/blob/master/visualization/include/pcl/visualization/impl/pcl_visualizer.hpp#L792-L794), it seems to conflict with your statement. It seems like the function is storing the start and end point of the line segment that will be drawn as a normal, and it's clearly lacking the prior transformation from the sensor orientation. 

Yet again, I might be misunderstanding what you're saying :smile:

**Comment by andersgb1** [comment]
Heh, @SergioRAgostinho,

Yes, it lacks the use of the sensor orientation (and origin), but that holds both for the point cloud and the normal cloud.

But, your code lines support my statement :) They show exactly what I'm trying to say, that XYZ information used for placing the normal lines is read out from the point cloud **cloud**, and not the normal cloud **normals**. Am I confusing something here?

**Comment by Mariusmssj** [comment]
I found this to be a common bug across the multiple parts of the library. I use IMU positional sensor data with my point clouds and when creating normals or getting ISS3D key-points, the sensor orientation and origin information is not passed along. 
I fixed this manually by adding:

  output.sensor_orientation_ = input_->sensor_orientation_;
  output.sensor_origin_ = input_->sensor_origin_;

to the iss_3d.hpp for keypoint creating and in normal_3d.hpp for normals.
